### ISSUE TITLE
no-unused-vars should ignore necessary variables in eslint comprehensions

### ISSUE THREAD

**Issue by gfarrell** [eslint/eslint#7220]
**Tell us about your environment**

* **ESLint Version:** v3.5.0
* **Node Version:** v6.2.1
* **npm Version:** 3.9.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
{
    "extends": "eslint:recommended",
    "env": {
        "node": true,
        "es6": true
    },
    "rules": {
        "no-multi-spaces": 0,
        "no-underscore-dangle": 0,
        "no-mixed-spaces-and-tabs": 2,
        "block-scoped-var": 2,
        "no-inner-declarations": 2,
        "valid-jsdoc": 1,
        "no-irregular-whitespace": 2,
        "spaced-comment": [1, "always"],
        "no-unused-vars": 2,
        "no-undef": 2,
        "comma-dangle": [2, "always-multiline"],
        "no-console": 0,
        "no-debugger": 2,
        "no-dupe-args": 2,
        "no-dupe-keys": 2,
        "eqeqeq": 2,
        "no-useless-call": 1,
        "strict": [2, "global"],
        "max-len": [2, { "code": 100, "ignoreUrls": true }],
        "semi": 2,
        "eol-last": 2,
        "quotes": [2, "double"]
    },
    "parserOptions": {
        "ecmaVersion": 6,
        "ecmaFeatures": {
            "globalReturn": true
        }
    }
}
```

**What did you do? Please include the actual source code causing the issue.**
```javascript
// ...
.map(([ _, callback ]) => callback)
// ...
```

**What did you expect to happen?**
eslint to recognise that the first variable (`_`) is necessary in order to get the second, used variable.

**What actually happened? Please include the actual, raw output from ESLint.**
`26:13  error  '_' is defined but never used          no-unused-vars`

**Comment by ilyavolodin** [comment]
ESLint does have an option for this, try changing your configuration to:
```js
"no-unused-vars": ["error", { "args": "after-used" }]
```

**Comment by gfarrell** [comment]
@ilyavolodin ok, thanks. I can't think of a case where this should not be the default behaviour, however. This variable is, effectively, in use.

**Comment by mysticatea** [comment]
You can remove the `_` safely since array destructurings allow holes.

```js
.map(([ , callback ]) => callback)
```

**Comment by vitorbal** [comment]
My guess is the behavior is an option and no the default to avoid breaking changes, but I could be wrong. I think @mysticatea's suggestion is also a good alternative. Closing, as the issue seems to be resolved. Please reopen if there is more to discuss.
